docs/source/extending/entrypoints.rst

```diff
@@ -0,0 +1,46 @@
+Registering Extensions with Entry Points
+========================================
+
+Often, third party packages will have a user-facing API as well as define
+extensions to the Numba compiler.  In those situations, the new types and
+overloads can registered with Numba when the package is imported by the user.
+However, there are situations where a Numba extension would not normally be
+imported directly by the user, but must still be registered with the Numba
+compiler.  An example of this is the `numba-scipy
+<https://github.com/numba/numba-scipy>`_ package, which adds support for some
+SciPy functions to Numba.  The end user does not need to ``import
+numba_scipy`` to enable compiler support for SciPy.  The extension only needs
+to be installed into the Python environment.
+
+Numba discovers extensions with the `entry points
+<https://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins>`_
+feature of ``setuptools``.  This allows a Python package to register an
+initializer function that will be called before ``numba`` compiles for the
+first time.  The delay ensures that the cost of importing extensions is
+deferred until it is necessary.
+
+
+Adding Support for the "Init" Entry Point
```

Comment:
Deferring review of this part until https://github.com/numba/numba/pull/4427#issuecomment-523226943 is resolved.

Reply:
I think we'll do this first PR using the init entry point.  While more fine grained entry points may be useful, we don't have the mechanism yet to trigger based on detected types.